File: app/src/main/kotlin/com/spacex/ui/RocketLaunchesApp.java
package com.spacex.ui;

import com.spacex.ui.di.AppComponent;
import com.squareup.leakcanary.LeakCanary;
import dagger.android.AndroidInjector;
import dagger.android.DaggerApplication;
import timber.log.Timber;

public class RocketLaunchesApp extends DaggerApplication {

  private static RocketLaunchesApp instance;

  public static RocketLaunchesApp getInstance() {
    return instance;
  }

  @Override
  protected AndroidInjector<? extends DaggerApplication> applicationInjector() {
    final AppComponent appComponent = DaggerAppComponent.builder().application(this).build();
    appComponent.inject(this);
    return appComponent;
  }

  @Override
  public void onCreate() {
    super.onCreate();
    instance = this;
    initLeakCanary();
    initTimber();
  }

  private void initLeakCanary() {
    if (LeakCanary.isInAnalyzerProcess(this)) {
      // This process is dedicated to LeakCanary for heap analysis.
      // You should not init your app in this process.
      return;
    }
    LeakCanary.install(this);
  }

  private void initTimber() {
    if (BuildConfig.DEBUG) {
      Timber.plant(new Timber.DebugTree());
    }
  }
}

File: app/src/main/kotlin/com/spacex/ui/RocketLaunchesApp.kt
package com.spacex.ui

import com.spacex.ui.di.DaggerAppComponent
import com.squareup.leakcanary.LeakCanary
import dagger.android.AndroidInjector
import dagger.android.DaggerApplication
import timber.log.Timber

class RocketLaunchesApp : DaggerApplication() {

    companion object {
        lateinit var instance: RocketLaunchesApp
            private set
    }

    override fun applicationInjector(): AndroidInjector<out DaggerApplication> {
        val appComponent = DaggerAppComponent.builder().application(this).build()
        appComponent.inject(this)
        return appComponent
    }

    override fun onCreate() {
        super.onCreate()
        instance = this
        initLeakCanary()
        initTimber()
    }

    private fun initLeakCanary() {
        if (LeakCanary.isInAnalyzerProcess(this)) {
            // This process is dedicated to LeakCanary for heap analysis.
            // You should not init your app in this process.
            return
        }
        LeakCanary.install(this)
    }

    private fun initTimber() {
        if (BuildConfig.DEBUG) {
            Timber.plant(Timber.DebugTree())
        }
    }
}
